### PR TITLE
Refactoring of TreeGrid

### PR DESCRIPTION
**Pull request by ashklianko**

### CODE REVIEW

File: src/main/resources/assets/admin/common/js/ui/treegrid/TreeGrid.ts [hunk 1]
```diff
@@ -2088,4 +1724,85 @@ export class TreeGrid<DATA>
...
+        root.setChildren(rootChildren);
+
+        return rootChildren.indexOf(item);
+    }
+
+    getDataByRow(row: number): DATA {
+        const nodes: TreeNode<DATA>[] = this.getRoot().getCurrentRoot().treeToList();
+        const node = nodes[row];
+
+        return !!node ? node.getData() : null;
```
Comment (jsi) [inline src/main/resources/assets/admin/common/js/ui/treegrid/TreeGrid.ts:1771]: Codacy found an issue: [Redundant double negation.](https://app.codacy.com/manual/enonic/lib-admin-ui/pullRequest?prid=6401650)

**Comment by jsi** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 18
- Added 16
           

Complexity increasing per file
==============================
- src/main/resources/assets/admin/common/js/ui/selector/Option.ts  1
         

Complexity decreasing per file
==============================
+ src/main/resources/assets/admin/common/js/ui/treegrid/TreeNode.ts  -2
         

Clones added
============
- src/main/resources/assets/admin/common/js/ui/treegrid/TreeNode.ts  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/enonic/lib-admin-ui/pullRequest?prid=6401650&bid=20752844)

File: src/main/resources/assets/admin/common/js/ui/treegrid/TreeGrid.ts [hunk 2]
```diff
@@ -627,231 +581,91 @@ export class TreeGrid<DATA>
...
-
-        return this.fetchAndUpdateNodes(nodesToUpdate, oldDataId ? this.getDataId(data) : undefined);
+        nodes.forEach((node: TreeNode<DATA>) => {
+            this.deleteNode(node);
+        });
     }
 
-    deleteNode(data: DATA): void {
-        if (this.highlightedNode && this.highlightedNode.getDataId() === this.getDataId(data)) {
+    protected deleteNode(node: TreeNode<DATA>): void {
```
Comment (jsi) [inline src/main/resources/assets/admin/common/js/ui/treegrid/TreeGrid.ts:640]: Codacy found an issue: [Method 'deleteNode' has a complexity of 7. Maximum allowed is 4.](https://app.codacy.com/manual/enonic/lib-admin-ui/pullRequest?prid=6401650)

File: src/main/resources/assets/admin/common/js/ui/treegrid/TreeGrid.ts [hunk 3]
```diff
@@ -2088,4 +1724,85 @@ export class TreeGrid<DATA>
...
+        const node: TreeNode<DATA> = this.root.getNodeByDataIdFromCurrent(dataId);
+
+        return !!node ? node.isExpandable() && node.hasChildren() : false;
+    }
+
+    protected isToBeExpanded(node: TreeNode<DATA>): boolean {
+        return this.expandedNodesDataIds.indexOf(node.getDataId()) > -1;
+    }
+
+    getDataFromDomEvent(e: DOMEvent): DATA {
```
Comment (jsi) [inline src/main/resources/assets/admin/common/js/ui/treegrid/TreeGrid.ts:1803]: Codacy found an issue: ['DOMEvent' is not defined.](https://app.codacy.com/manual/enonic/lib-admin-ui/pullRequest?prid=6401650)

File: src/main/resources/assets/admin/common/js/ui/treegrid/TreeGrid.ts [hunk 4]
```diff
@@ -2088,4 +1724,85 @@ export class TreeGrid<DATA>
...
+
+    getDataLevel(data: DATA): number {
+        const node: TreeNode<DATA> = this.root.getNodeByDataIdFromCurrent(data.getId());
+        return !!node ? node.calcLevel() : -1;
+    }
+
+    isExpandedAndHasChildren(dataId: string): boolean {
+        const node: TreeNode<DATA> = this.root.getNodeByDataIdFromCurrent(dataId);
+
+        return !!node ? node.isExpandable() && node.hasChildren() : false;
```
Comment (jsi) [inline src/main/resources/assets/admin/common/js/ui/treegrid/TreeGrid.ts:1796]: Codacy found an issue: [Redundant double negation.](https://app.codacy.com/manual/enonic/lib-admin-ui/pullRequest?prid=6401650)

File: src/main/resources/assets/admin/common/js/ui/treegrid/TreeGrid.ts [hunk 5]
```diff
@@ -245,31 +251,31 @@ export class TreeGrid<DATA>
...
     }
 
-    onHighlightingChanged(listener: (node: TreeNode<DATA>, force: boolean, callback: Function) => void) {
+    onHighlightingChanged(listener: () => void) {
         this.highlightingChangeListeners.push(listener);
         return this;
     }
 
-    public getRowByNode(node: TreeNode<DATA>): JQuery {
+    private getRowByNode(node: TreeNode<DATA>): JQuery {
```
Comment (jsi) [inline src/main/resources/assets/admin/common/js/ui/treegrid/TreeGrid.ts:259]: Codacy found an issue: ['JQuery' is not defined.](https://app.codacy.com/manual/enonic/lib-admin-ui/pullRequest?prid=6401650)

File: src/main/resources/assets/admin/common/js/ui/treegrid/TreeGrid.ts [hunk 6]
```diff
@@ -2012,21 +1650,21 @@ export class TreeGrid<DATA>
         this.setActive(true);
     }
 
-    private highlightRowByNode(node: TreeNode<DATA>, immediateNotification: boolean = false, callback?: Function) {
+    private highlightRowByNode(node: TreeNode<DATA>) {
```
Comment (jsi) [inline src/main/resources/assets/admin/common/js/ui/treegrid/TreeGrid.ts:1653]: Codacy found an issue: [Method 'highlightRowByNode' has a complexity of 7. Maximum allowed is 4.](https://app.codacy.com/manual/enonic/lib-admin-ui/pullRequest?prid=6401650)

File: src/main/resources/assets/admin/common/js/ui/selector/Option.ts [hunk 7]
```diff
@@ -1,5 +1,82 @@
...
+    isSelectable(): boolean {
+        return this.selectable;
+    }
+
+    isExpandable(): boolean {
+        return this.expandable;
+    }
+
+    static create<T>(): OptionBuilder<T> {
+        return new OptionBuilder<T>();
```
Comment (jsi) [inline src/main/resources/assets/admin/common/js/ui/selector/Option.ts:71]: Codacy found an issue: ['OptionBuilder' was used before it was defined.](https://app.codacy.com/manual/enonic/lib-admin-ui/pullRequest?prid=6401650)

File: src/main/resources/assets/admin/common/js/ui/selector/Option.ts [hunk 8]
```diff
@@ -1,5 +1,82 @@
...
+
+    private readOnly?: boolean;
+
+    private empty?: boolean;
+
+    private selectable?: boolean;
+
+    private expandable?: boolean;
+
+    constructor(builder: OptionBuilder<T>) {
```
Comment (jsi) [inline src/main/resources/assets/admin/common/js/ui/selector/Option.ts:20]: Codacy found an issue: ['OptionBuilder' was used before it was defined.](https://app.codacy.com/manual/enonic/lib-admin-ui/pullRequest?prid=6401650)

File: src/main/resources/assets/admin/common/js/ui/selector/Option.ts [hunk 9]
```diff
@@ -1,5 +1,82 @@
...
+
+    isSelectable(): boolean {
+        return this.selectable;
+    }
+
+    isExpandable(): boolean {
+        return this.expandable;
+    }
+
+    static create<T>(): OptionBuilder<T> {
```
Comment (jsi) [inline src/main/resources/assets/admin/common/js/ui/selector/Option.ts:70]: Codacy found an issue: ['OptionBuilder' was used before it was defined.](https://app.codacy.com/manual/enonic/lib-admin-ui/pullRequest?prid=6401650)

File: src/main/resources/assets/admin/common/js/ui/treegrid/TreeGrid.ts [hunk 10]
```diff
@@ -627,231 +581,91 @@ export class TreeGrid<DATA>
...
     }
 
-    updateNodeByDataId(dataId: string): Q.Promise<void> {
-        const nodesToUpdate: TreeNode<DATA>[] = this.root.getNodesByDataId(dataId);
-
-        if (nodesToUpdate.length > 0) {
-            return this.fetchAndUpdateNodes(nodesToUpdate);
-        }
+    updateNodeByData(updatedData: DATA, oldDataId?: string) {
+        const dataId: string = !!oldDataId ? oldDataId : updatedData.getId();
```
Comment (jsi) [inline src/main/resources/assets/admin/common/js/ui/treegrid/TreeGrid.ts:622]: Codacy found an issue: [Redundant double negation.](https://app.codacy.com/manual/enonic/lib-admin-ui/pullRequest?prid=6401650)

File: src/main/resources/assets/admin/common/js/ui/treegrid/TreeGrid.ts [hunk 11]
```diff
@@ -2088,4 +1724,85 @@ export class TreeGrid<DATA>
...
+
+        return !!node ? node.isExpandable() && node.hasChildren() : false;
+    }
+
+    protected isToBeExpanded(node: TreeNode<DATA>): boolean {
+        return this.expandedNodesDataIds.indexOf(node.getDataId()) > -1;
+    }
+
+    getDataFromDomEvent(e: DOMEvent): DATA {
+        const cell: Slick.Cell = this.grid.getCellFromEvent(e);
```
Comment (jsi) [inline src/main/resources/assets/admin/common/js/ui/treegrid/TreeGrid.ts:1804]: Codacy found an issue: ['Slick' is not defined.](https://app.codacy.com/manual/enonic/lib-admin-ui/pullRequest?prid=6401650)

File: src/main/resources/assets/admin/common/js/ui/treegrid/TreeGrid.ts [hunk 12]
```diff
@@ -2088,4 +1724,85 @@ export class TreeGrid<DATA>
...
+        return node.getParent().getData();
+    }
+
+    getCurrentData(expanded: boolean = true): DATA[] {
+        return this.root.getCurrentRoot().treeToList(false, expanded).map((node: TreeNode<DATA>) => node.getData());
+    }
+
+    getDataLevel(data: DATA): number {
+        const node: TreeNode<DATA> = this.root.getNodeByDataIdFromCurrent(data.getId());
+        return !!node ? node.calcLevel() : -1;
```
Comment (jsi) [inline src/main/resources/assets/admin/common/js/ui/treegrid/TreeGrid.ts:1790]: Codacy found an issue: [Redundant double negation.](https://app.codacy.com/manual/enonic/lib-admin-ui/pullRequest?prid=6401650)

File: src/main/resources/assets/admin/common/js/ui/treegrid/TreeGrid.ts [hunk 13]
```diff
@@ -2088,4 +1724,85 @@ export class TreeGrid<DATA>
...
+
+        this.initData(rootChildren);
+        root.setChildren(rootChildren);
+
+        return rootChildren.indexOf(item);
+    }
+
+    getDataByRow(row: number): DATA {
+        const nodes: TreeNode<DATA>[] = this.getRoot().getCurrentRoot().treeToList();
+        const node = nodes[row];
```
Comment (jsi) [inline src/main/resources/assets/admin/common/js/ui/treegrid/TreeGrid.ts:1769]: Codacy found an issue: [Variable Assigned to Object Injection Sink](https://app.codacy.com/manual/enonic/lib-admin-ui/pullRequest?prid=6401650)

File: src/main/resources/assets/admin/common/js/ui/selector/Option.ts [hunk 14]
```diff
@@ -1,5 +1,82 @@
-export interface Option<T>
-    extends Slick.SlickData {
+import {IDentifiable} from '../../IDentifiable';
+
+export class Option<T>
+    implements Slick.SlickData, IDentifiable {
```
Comment (jsi) [inline src/main/resources/assets/admin/common/js/ui/selector/Option.ts:4]: Codacy found an issue: ['Slick' is not defined.](https://app.codacy.com/manual/enonic/lib-admin-ui/pullRequest?prid=6401650)

File: src/main/resources/assets/admin/common/js/ui/treegrid/TreeGrid.ts [hunk 15]
```diff
@@ -956,15 +770,21 @@ export class TreeGrid<DATA>
         return this.grid.isAnySelected();
     }
 
+    collapseNodeByRow(row: number) {
+        const nodes: TreeNode<DATA>[] = this.getRoot().getCurrentRoot().treeToList();
+        const draggedNode: TreeNode<DATA> = nodes[row];
```
Comment (jsi) [inline src/main/resources/assets/admin/common/js/ui/treegrid/TreeGrid.ts:775]: Codacy found an issue: [Variable Assigned to Object Injection Sink](https://app.codacy.com/manual/enonic/lib-admin-ui/pullRequest?prid=6401650)

File: src/main/resources/assets/admin/common/js/dom/ElementHelper.ts [hunk 16]
```diff
@@ -764,8 +764,8 @@ export class ElementHelper {
     }
 
     getChildren(): Node[] {
-
-        return this.el.children || //children property not supported for IE SVGelement, Document and DocumentFragment
+        //children property not supported for IE SVGelement, Document and DocumentFragment
+        return !!this.el.children ? [].slice.call(this.el.children) :
```
Comment (jsi) [inline src/main/resources/assets/admin/common/js/dom/ElementHelper.ts:768]: Codacy found an issue: [Redundant double negation.](https://app.codacy.com/manual/enonic/lib-admin-ui/pullRequest?prid=6401650)

**Review by alansemenov — APPROVED** [review]

**Pull request merged**